staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -851,14 +851,14 @@ func (c *Cacher) dispatchEvent(event *watchCacheEvent) {
 		// from it justifies increased memory usage, so for now we drop the cached
 		// serializations after dispatching this event.
 		//
-		// Given the deep-copies that are done to create cachingObjects,
-		// we try to cache serializations only if there are at least 3 watchers.
-		if len(c.watchersBuffer) >= 3 {
-			// Make a shallow copy to allow overwriting Object and PrevObject.
-			wcEvent := *event
-			setCachingObjects(&wcEvent, c.versioner)
-			event = &wcEvent
-		}
+		// Given that CachingObject is just wrapping the object and not perfoming
+		// deep-copying (until some field is explicitly being modified), we create
+		// it unconditionally to ensure safety and reduce deep-copying.
```

Comment:
it looks like cacheableObject still deep-copies at encode time, so this isn't yet saving deepcopies for any event that gets encoded, right?

are there other downsides to using cacheable object like increased memory use of the cached encoding that should keep us from expanding use unconditionally?

Reply:
> it looks like cacheableObject still deep-copies at encode time, so this isn't yet saving deepcopies for any event that gets encoded, right?

It does. But before we were always doing additional copy at the beginning, which we're no longer doing now.
So before this change what we were doing was:
- deepCopy during CachingObject creation
- deepCopy for every unique encoding
With this change - it's only the latter.

I would eventually like to get rid of that too, but I'm not yet convinced if it's safe, so one step at a time.

> are there other downsides to using cacheable object like increased memory use of the cached encoding that should keep us from expanding use unconditionally?

This is now doing it unconditionally, right?

Reply:
> This is now doing it unconditionally, right?

Yeah, I'm asking if the increased cost of caching the encoding is ok to bear all the time

Reply:
> Yeah, I'm asking if the increased cost of caching the encoding is ok to bear all the time

yes - it's fine, because without the initial deep-copy, even with just a single watcher we're roughly on-par in both cases